atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java

```diff
@@ -336,11 +345,82 @@ public void hostIsAutomaticallyAddedOnRefresh() {
         setCassandraServersTo(CASS_SERVER_1, CASS_SERVER_2);
 
         createClientPool();
-        assertThat(poolServers).containsExactlyInAnyOrder(CASS_SERVER_1, CASS_SERVER_2);
+        assertThatMap(poolServers).containsOnlyKeys(CASS_SERVER_1, CASS_SERVER_2);
 
         setCassandraServersTo(CASS_SERVER_1, CASS_SERVER_2, CASS_SERVER_3);
-        deterministicExecutor.tick(POOL_REFRESH_INTERVAL_SECONDS, TimeUnit.SECONDS);
-        assertThat(poolServers).containsExactlyInAnyOrder(CASS_SERVER_1, CASS_SERVER_2, CASS_SERVER_3);
+        refreshPool();
+        assertThatMap(poolServers).containsOnlyKeys(CASS_SERVER_1, CASS_SERVER_2, CASS_SERVER_3);
+    }
+
+    @Test
+    public void onlyValidatedHostsAreAddedOnRefresh() {
+        setupThriftServers(ImmutableSet.of(CASS_SERVER_1.proxy()));
+        when(config.autoRefreshNodes()).thenReturn(true);
+
+        setCassandraServersTo(CASS_SERVER_1);
+
+        createClientPool();
+        assertThatMap(poolServers).containsOnlyKeys(CASS_SERVER_1);
+        setupHostsWithInconsistentTopology(CASS_SERVER_2);
+
+        setCassandraServersTo(CASS_SERVER_1, CASS_SERVER_2, CASS_SERVER_3);
+        refreshPool();
+        assertThatMap(poolServers).containsOnlyKeys(CASS_SERVER_1, CASS_SERVER_3);
+        assertThat(absentHostTracker.returnPool(CASS_SERVER_2)).isPresent();
+    }
+
+    @Test
+    public void hostsWithDifferingTopologiesTrackedAsAbsentHost() {
+        setupThriftServers(ImmutableSet.of(CASS_SERVER_1.proxy()));
+        when(config.autoRefreshNodes()).thenReturn(true);
+        setCassandraServersTo(CASS_SERVER_1, CASS_SERVER_2, CASS_SERVER_3);
+        setupHostsWithInconsistentTopology(CASS_SERVER_2);
+        createClientPool();
+        assertThat(absentHostTracker.returnPool(CASS_SERVER_2)).isPresent();
+    }
+
+    @Test
+    public void throwsWhenNoServersInPoolAndServersAddedPresentButAreInvalid() {
+        setupHostsWithInconsistentTopology(CASS_SERVER_1);
+        setupThriftServers(ImmutableSet.of(CASS_SERVER_1.proxy()));
+        when(config.autoRefreshNodes()).thenReturn(true);
+        setCassandraServersTo(CASS_SERVER_1);
+        assertThatThrownBy(this::createClientPool).isInstanceOf(SafeIllegalStateException.class);
```

Comment:
I think Jeremy mentioned this in the last PR, but we generally test the exception message and args too :)
I'd recommend using assertThatLoggableExceptionThrownBy, and verifying the message and args are what you expect.